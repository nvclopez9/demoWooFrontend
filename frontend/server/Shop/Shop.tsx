import {
  Product,
  ProductCategory,
  PaColor,
} from '@woographql/graphql';
import { ShopCategories } from "@woographql/client/ShopCategories";
import { PaColorPicker } from '@woographql/client/PaColorPicker';
import { SearchBar } from '@woographql/client/SearchBar';
import { ProductListing } from '@woographql/client/ProductListing';
import { PriceRange } from '@woographql/client/PriceRange';
import { ShopSidebar } from '@woographql/server/ShopSidebar';

export interface ShopProps {
  products: Product[];
  categories?: ProductCategory[];
  colors: PaColor[];
}

export function Shop(props: ShopProps) {
  const {
    products,
    categories,
    colors,
  } = props;

  return (
    <div className="w-full flex max-w-screen-lg mx-auto">
      <ShopSidebar>
        {categories && (
          <>
            <p className="font-serif text-lg font-bold mb-2">Categories</p>
            <ShopCategories categories={categories} />
          </>
        )}
        <p className="font-serif text-lg font-bold mb-2">Colors</p>
        <PaColorPicker colors={colors} />
        <p className="font-serif text-lg font-bold mb-2">Price</p>
        <PriceRange />
      </ShopSidebar>
      <div className="w-full px-4 lg:w-3/4">
        <p className="font-serif text-lg font-bold mb-2">Search</p>
        <SearchBar />
        <p className="font-serif text-lg font-bold mb-2">Results</p>
        <ProductListing products={products} />
      </div>
    </div>
  );
}
