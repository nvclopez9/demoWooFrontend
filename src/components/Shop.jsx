import React, { useState, useEffect } from "react";
import CartHelper from "../helpers/CartHelper";

const ProductsListComponent = () => {
  const [products, setProducts] = useState([]);
  const [loading, setLoading] = useState(true);
  const [cartKey, setCartKey] = useState(CartHelper.getCartKey());
  const [cartMessage, setCartMessage] = useState(null);
  const [cartError, setCartError] = useState(null);
  const baseUrl = import.meta.env.PUBLIC_API_URL;

  // Cargar productos al inicio
  useEffect(() => {
    const fetchProducts = async () => {
      try {
        const response = await fetch(`${baseUrl}/cocart/v2/products`, {
          method: "GET",
          credentials: "include", // Enviar credenciales con la solicitud
        });
        if (!response.ok) {
          throw new Error("Failed to fetch products");
        }
        const data = await response.json();
        setProducts(data.products || []);
      } catch (error) {
        console.error("Error fetching products:", error.message);
      } finally {
        setLoading(false);
      }
    };

    fetchProducts();
  }, [baseUrl]);

  const addToCart = async (productId) => {
    try {
      // Construir la URL con el `cart_key` como par√°metro si existe
      let url = `${baseUrl}/cocart/v2/cart/add-item?id=${productId}&quantity=1`;
      const currentCartKey = CartHelper.getCartKey();

      if (currentCartKey) {
        url += `&cart_key=${currentCartKey}`;
      }

      const response = await fetch(url, {
        method: "POST",
        credentials: "include", // Enviar credenciales con la solicitud
      });

      if (!response.ok) {
        const errorData = await response.json();
        throw new Error(errorData.message || "Failed to add product to cart");
      }

      const data = await response.json();

      // Si no hay cartKey, guardar el que devuelve la API
      if (!currentCartKey && data.cart_key) {
        CartHelper.setCartKey(data.cart_key); // Guardar en la cookie
        setCartKey(data.cart_key);
      }

      setCartMessage("Product added to cart successfully!");
      console.log("Producto agregado correctamente:", data);
    } catch (error) {
      setCartError("An error occurred while adding the product.");
      console.error("Error adding product to cart:", error.message);
    }
  };

  return (
    <div className="min-h-screen bg-gray-50 flex flex-col items-center p-4">
      <div className="max-w-5xl w-full bg-white shadow-md rounded-lg p-6">
        <h1 className="text-3xl font-bold text-center mb-6">Products List</h1>

        {cartMessage && (
          <p className="text-center text-green-600 font-semibold mb-4">
            {cartMessage}
          </p>
        )}

        {cartError && (
          <p className="text-center text-red-600 font-semibold mb-4">
            {cartError}
          </p>
        )}

        {loading && (
          <p className="text-center text-gray-600">Loading products...</p>
        )}

        {!loading && products.length > 0 && (
          <div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-6">
            {products.map((product) => (
              <div
                key={product.id}
                className="border rounded-lg shadow p-4 bg-gray-100"
              >
                {product.images && product.images[0]?.src.full && (
                  <img
                    src={product.images[0].src.full}
                    alt={product.images[0].alt || product.name}
                    className="w-full h-48 object-cover mb-4 rounded"
                  />
                )}
                <h2 className="text-xl font-semibold text-gray-800 mb-2">
                  {product.name} - {product.prices.price / 100}{" "}
                  {product.prices.currency.currency_symbol}
                </h2>
                <p className="text-gray-700 mb-4">{product.description}</p>
                <button
                  onClick={() => addToCart(product.id)}
                  className="text-white bg-blue-500 hover:bg-blue-600 px-4 py-2 rounded"
                >
                  {product.add_to_cart?.text || "Add to Cart"}
                </button>
              </div>
            ))}
          </div>
        )}

        {!loading && products.length === 0 && (
          <p className="text-center text-gray-600">
            No products available at the moment.
          </p>
        )}
      </div>
    </div>
  );
};

export default ProductsListComponent;
