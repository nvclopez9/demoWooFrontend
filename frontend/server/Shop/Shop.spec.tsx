import '@testing-library/react/dont-cleanup-after-each';
import {
  products,
  categories,
  colors,
  render,
  screen,
} from '@woographql/testing';
import { Shop } from './Shop';

describe('Shop component', () => {
  const { baseElement } = render(
    <Shop products={products} categories={categories} colors={colors} />
  );

  it('renders correctly', async () => {
    // Check that the Categories, Colors, Price, Search, and Results headings are rendered
    expect(screen.getByText('Categories')).toBeInTheDocument();
    expect(screen.getByText('Colors')).toBeInTheDocument();
    expect(screen.getByText('Price')).toBeInTheDocument();
    expect(screen.getByText('Search')).toBeInTheDocument();
    expect(screen.getByText('Results')).toBeInTheDocument();

    // Check that the category names are rendered
    categories.forEach((category) => {
      expect(screen.getByText(category.name as string)).toBeInTheDocument();
    });

    // Check that the color names are rendered
    colors.forEach((color) => {
      expect(screen.getByText(color.name as string)).toBeInTheDocument();
    });

  // Check that the product names are rendered
    products.forEach((product) => {
      expect(screen.getByText(product.name as string)).toBeInTheDocument();
    });
  });

  it('should match snapshot', () => {
    expect(baseElement).toMatchSnapshot();
  });
});
