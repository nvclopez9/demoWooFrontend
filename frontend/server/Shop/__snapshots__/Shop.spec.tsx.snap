// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shop component should match snapshot 1`] = `
<body>
  <div>
    <div
      class="w-full flex max-w-screen-lg mx-auto"
    >
      <button
        aria-controls="radix-:r0:"
        aria-expanded="false"
        aria-haspopup="dialog"
        class="inline-flex items-center justify-center rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2 lg:hidden w-16 fixed inset-x-0 mx-auto bottom-10 z-30"
        data-state="closed"
        type="button"
      >
        <i
          aria-label="Open sidebar"
          class="fa-solid fa-filter"
        />
      </button>
      <div
        class="hidden lg:block w-1/4 px-4"
      >
        <p
          class="font-serif text-lg font-bold mb-2"
        >
          Categories
        </p>
        <div
          class="flex gap-2 flex-wrap mb-4"
        />
        <ul
          class="mb-4 max-h-[40vh] lg:max-h-[25vh] overflow-y-scroll scrollbar-thin scrollbar-corner-rounded scrollbar-thumb-ring"
        >
          <li
            class="group py-2"
          >
            <a
              class="transition-colors group-hover:text-blue-400"
              href=""
            >
              Clothing
            </a>
          </li>
          <li
            class="group py-2"
          >
            <a
              class="transition-colors group-hover:text-blue-400"
              href=""
            >
              Accessories
            </a>
          </li>
        </ul>
        <p
          class="font-serif text-lg font-bold mb-2"
        >
          Colors
        </p>
        <div
          class="flex gap-2 flex-wrap mb-4"
        />
        <ul
          class="mb-4 max-h-[40vh] lg:max-h-[25vh] overflow-y-scroll scrollbar-thin scrollbar-corner-rounded scrollbar-thumb-ring"
        >
          <li
            class="group py-4"
          >
            <a
              class="px-0 flex gap-2 transition-colors group-hover:text-blue-400"
              href=""
            >
              <div
                class="rounded w-6 h-6"
                style="background-color: red;"
              />
              Red
            </a>
          </li>
          <li
            class="group py-4"
          >
            <a
              class="px-0 flex gap-2 transition-colors group-hover:text-blue-400"
              href=""
            >
              <div
                class="rounded w-6 h-6"
                style="background-color: blue;"
              />
              Blue
            </a>
          </li>
          <li
            class="group py-4"
          >
            <a
              class="px-0 flex gap-2 transition-colors group-hover:text-blue-400"
              href=""
            >
              <div
                class="rounded w-6 h-6"
                style="background-color: green;"
              />
              Green
            </a>
          </li>
          <li
            class="group py-4"
          >
            <a
              class="px-0 flex gap-2 transition-colors group-hover:text-blue-400"
              href=""
            >
              <div
                class="rounded w-6 h-6"
                style="background-color: yellow;"
              />
              Yellow
            </a>
          </li>
        </ul>
        <p
          class="font-serif text-lg font-bold mb-2"
        >
          Price
        </p>
        <div
          class="mb-4 flex gap-x-2 gap-y-4 flex-wrap"
        >
          <input
            class="flex h-9 w-full rounded-md border border-input bg-background px-3 py-1 text-sm shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 basis-1/2 shrink"
            max="100"
            min="0"
            type="number"
            value="0"
          />
          <input
            class="flex h-9 w-full rounded-md border border-input bg-background px-3 py-1 text-sm shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 basis-0 grow"
            max="100"
            min="0"
            type="number"
            value="100"
          />
          <span
            aria-disabled="false"
            class="relative flex touch-none select-none items-center w-full"
            data-orientation="horizontal"
            dir="ltr"
            style="--radix-slider-thumb-transform: translateX(-50%);"
          >
            <span
              class="relative h-1.5 w-full grow overflow-hidden rounded-full bg-primary/20"
              data-orientation="horizontal"
            >
              <span
                class="absolute h-full bg-primary"
                data-orientation="horizontal"
                style="left: 0%; right: 0%;"
              />
            </span>
            <span
              style="transform: var(--radix-slider-thumb-transform); position: absolute; left: calc(0% + 0px);"
            >
              <span
                aria-label="Minimum"
                aria-orientation="horizontal"
                aria-valuemax="100"
                aria-valuemin="0"
                aria-valuenow="0"
                class="block h-4 w-4 rounded-full border border-primary/50 bg-background shadow transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50"
                data-orientation="horizontal"
                data-radix-collection-item=""
                role="slider"
                style=""
                tabindex="0"
              />
            </span>
            <span
              style="transform: var(--radix-slider-thumb-transform); position: absolute; left: calc(100% + 0px);"
            >
              <span
                aria-label="Maximum"
                aria-orientation="horizontal"
                aria-valuemax="100"
                aria-valuemin="0"
                aria-valuenow="100"
                class="block h-4 w-4 rounded-full border border-primary/50 bg-background shadow transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50"
                data-orientation="horizontal"
                data-radix-collection-item=""
                role="slider"
                style=""
                tabindex="0"
              />
            </span>
          </span>
        </div>
      </div>
      <div
        class="w-full px-4 lg:w-3/4"
      >
        <p
          class="font-serif text-lg font-bold mb-2"
        >
          Search
        </p>
        <input
          class="flex h-9 w-full rounded-md border border-input bg-background px-3 py-1 text-sm shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 mb-4"
          value=""
        />
        <p
          class="font-serif text-lg font-bold mb-2"
        >
          Results
        </p>
        <p
          class="font-semibold mb-4"
        >
          Showing 
          4
           of 
          4
           items
        </p>
        <div
          class="flex flex-wrap justify-center md:justify-start gap-4"
        >
          <a
            href="/product/t-shirt"
          >
            <div
              class="rounded-xl border bg-card text-card-foreground shadow w-44 md:w-36"
            >
              <div
                class="flex flex-col space-y-1.5 p-4"
              >
                <h3
                  class="font-semibold leading-none tracking-tight font-serif whitespace-nowrap truncate w-full"
                >
                  T Shirt
                </h3>
              </div>
              <div
                class="p-4"
              >
                <p
                  class="text-sm truncate"
                />
              </div>
              <div
                class="flex items-center p-4"
              >
                <p
                  class="font-serif font-bold"
                >
                  $10.00
                </p>
              </div>
            </div>
          </a>
          <a
            href="/product/belt"
          >
            <div
              class="rounded-xl border bg-card text-card-foreground shadow w-44 md:w-36"
            >
              <div
                class="flex flex-col space-y-1.5 p-4"
              >
                <h3
                  class="font-semibold leading-none tracking-tight font-serif whitespace-nowrap truncate w-full"
                >
                  Belt
                </h3>
              </div>
              <div
                class="p-4"
              >
                <p
                  class="text-sm truncate"
                />
              </div>
              <div
                class="flex items-center p-4"
              >
                <p
                  class="font-serif font-bold"
                >
                  $20.00
                </p>
              </div>
            </div>
          </a>
          <a
            href="/product/shoes"
          >
            <div
              class="rounded-xl border bg-card text-card-foreground shadow w-44 md:w-36"
            >
              <div
                class="flex flex-col space-y-1.5 p-4"
              >
                <h3
                  class="font-semibold leading-none tracking-tight font-serif whitespace-nowrap truncate w-full"
                >
                  Shoes
                </h3>
              </div>
              <div
                class="p-4"
              >
                <p
                  class="text-sm truncate"
                />
              </div>
              <div
                class="flex items-center p-4"
              >
                <p
                  class="font-serif font-bold"
                >
                  $30.00
                </p>
              </div>
            </div>
          </a>
          <a
            href="/product/hat"
          >
            <div
              class="rounded-xl border bg-card text-card-foreground shadow w-44 md:w-36"
            >
              <div
                class="flex flex-col space-y-1.5 p-4"
              >
                <h3
                  class="font-semibold leading-none tracking-tight font-serif whitespace-nowrap truncate w-full"
                >
                  Hat
                </h3>
              </div>
              <div
                class="p-4"
              >
                <p
                  class="text-sm truncate"
                />
              </div>
              <div
                class="flex items-center p-4"
              >
                <p
                  class="font-serif font-bold"
                >
                  $40.00
                </p>
              </div>
            </div>
          </a>
        </div>
        <div
          class="flex justify-center my-4 gap-x-2 text-sms"
        >
          <a
            aria-label="Previous page"
            class="text-gray-400 opacity-50 pointer-events-none self-center rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring bg-primary shadow hover:bg-primary/90 h-9 px-4 py-2"
            href=""
            role="button"
          >
            <i
              aria-hidden="true"
              class="fa-solid fa-chevron-left leading-[0.95]"
            />
          </a>
          <div
            class="flex flex-wrap justify-center gap-2"
          />
          <a
            aria-label="Next page"
            class="text-gray-400 opacity-50 pointer-events-none self-center rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring bg-primary shadow hover:bg-primary/90 h-9 px-4 py-2"
            href=""
            role="button"
          >
            <i
              aria-hidden="true"
              class="fa-solid fa-chevron-right"
            />
          </a>
        </div>
      </div>
    </div>
  </div>
</body>
`;
